authorizer-app-backend/src/main/java/org/radarbase/authorizer/resources/RestSourceUserResource.kt

```diff
@@ -0,0 +1,154 @@
+import org.radarbase.jersey.auth.Auth
+import org.radarbase.jersey.auth.Authenticated
+import org.radarbase.jersey.exception.HttpBadRequestException
+import org.radarbase.jersey.exception.HttpConflictException
+import org.radarbase.jersey.exception.HttpNotFoundException
+import org.radarcns.auth.authorization.Permission
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.net.URI
+import javax.annotation.Resource
+import javax.inject.Singleton
+import javax.ws.rs.*
+import javax.ws.rs.core.Context
+import javax.ws.rs.core.MediaType
+import javax.ws.rs.core.Response
+
+@Path("users")
+@Produces(MediaType.APPLICATION_JSON)
+@Consumes(MediaType.APPLICATION_JSON)
+@Resource
+@Authenticated
+@Singleton
+class RestSourceUserResource(
+    @Context private val userRepository: RestSourceUserRepository,
+    @Context private val userMapper: RestSourceUserMapper,
+    @Context private val authorizationService: RestSourceAuthorizationService,
+    @Context private val projectService: RadarProjectService,
+    @Context private val auth: Auth
+) {
+
+    @GET
+    fun query(
+        @QueryParam("projectId") projectId: String?,
+        @QueryParam("sourceType") sourceType: String?,
+        @QueryParam("size") pageSize: Int?,
+        @DefaultValue("1") @QueryParam("page") pageNumber: Int): RestSourceUsers {
+
+        if (projectId != null) {
+            auth.checkPermissionOnProject(Permission.PROJECT_READ, projectId)
+        }
+
+        val queryPage = Page(pageNumber = pageNumber, pageSize = pageSize)
+        val (records, page) = userRepository.query(queryPage, projectId, sourceType)
+
+        return userMapper.fromRestSourceUsers(records, page)
+    }
+
+    @POST
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    fun create(
+        @FormParam("code") code: String,
+        @FormParam("state") state: String): Response {
+        logger.info("code $code state $state")
+        val accessToken = authorizationService.requestAccessToken(code, sourceType = state)
+        val user = userRepository.createOrUpdate(accessToken, state)
+
+        return Response.created(URI("users/${user.id}"))
+            .entity(userMapper.fromEntity(user))
+            .build()
+    }
+
+    @POST
+    @Path("{id}")
+    fun update(
+        @PathParam("id") userId: Long,
+        user: RestSourceUserDTO): RestSourceUserDTO {
+        val existingUser = validate(userId, user, Permission.SUBJECT_UPDATE)
+
+        val updatedUser = userRepository.update(existingUser, user)
+        return userMapper.fromEntity(updatedUser)
+    }
+
+    @GET
+    @Path("{id}")
+    fun readUser(@PathParam("id") userId: Long): RestSourceUserDTO {
+        val user = ensureUser(userId)
+        auth.checkPermissionOnSubject(Permission.SUBJECT_READ, user.projectId, user.userId)
+        return userMapper.fromEntity(user)
+    }
+
+    @DELETE
+    @Path("{id}")
+    fun deleteUser(@PathParam("id") userId: Long): Response {
```

Comment:
I'm thinking now that before any database request is made, the SUBJECT_READ scope should be checked for being present. Either with a `@NeedsPermission` or with an `auth.checkPermission` call in `ensureUser`. This avoids unauthorised users to make database requests.

Reply:
Am I right to understand that NeedsPermission only check for permission, but no dependency on subject or project. Then yes. Its better to add the check for permission.

Reply:
Indeed, it can only check for project / subject if that is included in the path. But if you leave them out, it just calls `auth.checkPermission`